questionpy_sdk/package/builder.py

```diff
@@ -126,6 +126,116 @@ def _run_hook(self, cmd: str, hook_name: BuildHookName, num: int) -> None:
+        pass
+
+    def __exit__(
+        self, exc_type: type[BaseException] | None, exc_value: BaseException | None, traceback: TracebackType | None
+    ) -> None:
+        return None
+
+
+class DirPackageBuilder(PackageBuilderBase):
+    """Builds a package to a `dist` directory."""
+
+    def __init__(self, source: PackageSource):
+        """Opens a source directory for writing.
+
+        Args:
+            source: Package source.
+        """
+        PackageBuilderBase.__init__(self, source, backup=False)
+
+    def _write_file(self, source_path: Path, dest_path: Path) -> None:
+        if not source_path.is_dir():
+            abs_dest_path = self._source.path / dest_path
+            self._ensure_target_dir(abs_dest_path)
+            shutil.copy(source_path, abs_dest_path)
+
+    def _write_string(self, dest_path: Path, content: str) -> None:
+        abs_dest_path = self._source.path / dest_path
+        self._ensure_target_dir(abs_dest_path)
+        with (abs_dest_path).open("w") as f:
+            f.write(content)
+
+    def _ensure_target_dir(self, path: Path) -> None:
+        makedirs(path.parent, exist_ok=True)
+
+
+class ZipPackageBuilder(PackageBuilderBase):
+    """Builds `.qpy` package file."""
+
+    COMPRESS_TYPE = zipfile.ZIP_DEFLATED
+
+    def __init__(self, output_path: PathLike, source: PackageSource, *, backup: bool = True):
+        """Opens a package file for writing.
+
+        Args:
+            output_path: Package output path.
+            source: Package source.
+            backup: Create source backup inside archive.
+        """
+        PackageBuilderBase.__init__(self, source, backup=backup)
+        self._zipfile = zipfile.ZipFile(output_path, mode="w", compression=self.COMPRESS_TYPE)
+
+    def _write_file(self, source_path: Path, dest_path: Path) -> None:
+        self._ensure_directory_entries(dest_path)
+        self._zipfile.write(source_path, dest_path, compress_type=self.COMPRESS_TYPE)
+
+    def _write_string(self, dest_path: Path, content: str) -> None:
+        self._ensure_directory_entries(dest_path)
+        self._zipfile.writestr(str(dest_path), content, compress_type=self.COMPRESS_TYPE)
+
+    def _ensure_directory_entries(self, path: Path) -> None:
+        """Ensure directory entries up to `path` are created."""
+        tz = datetime.datetime.now().astimezone().tzinfo
+        for parent in reversed(path.parents):
+            strpath = str(parent)
+            if not strpath.endswith("/"):
+                strpath += "/"
+            if len(parent.parts) > 0 and all(strpath != fn.filename for fn in self._zipfile.infolist()):
+                zipinfo = zipfile.ZipInfo(strpath, date_time=datetime.datetime.now(tz).timetuple()[:6])
+                zipinfo.compress_type = self.COMPRESS_TYPE
+                zipinfo.CRC = 0  # TODO: remove once bug is resolved (https://github.com/python/cpython/issues/119052)
+                self._zipfile.mkdir(zipinfo)
```

Comment:
Verstehe ich das richtig, dass wir `mkdir(str)` nur nicht nutzen, um den `compress_type` setzen zu können? 

Aber ZIP komprimiert ja immer nur auf Dateiebene, und ein Verzeichnis hat Größe 0. Dann ist Komprimierung hier doch irrelevant, und wir können den Block erstzen:
```suggestion
        for parent in reversed(path.parents):
            strpath = str(parent)
            if not strpath.endswith("/"):
                strpath += "/"
            if len(parent.parts) > 0 and strpath not in self._zipfile.namelist():
                self._zipfile.mkdir(strpath)
```

Reply:
> Verstehe ich das richtig, dass wir `mkdir(str)` nur nicht nutzen, um den `compress_type` setzen zu können?

Nein, geht nicht um `compress_type`. Geht um das Datum. `mkdir(str)` führt dazu, dass alle entries den timestamp `0` (01.01.1970) erhalten. Ist eher ein Detail, aber korrekter so, denke ich.

`compress_type` setze ich nur damit es konsistent mit dem Rest der zip entries ist, macht aber keinen Unterschied.

Reply:
Crap dass das nötig ist, ich war davon ausgegangen, dass der aktuelle Zeitpunkt der Default ist. Dann ist es fein.